### ISSUE TITLE
New Rule: no-ms-high-contrast

### ISSUE THREAD

**Issue by NotWoods** [eslint/css#17]
### Rule details

Auto fix uses of -ms-high-constrast to forced-colors

### What type of rule is this?

Warns about a potential problem

### Example code

```css
-ms-high-contrast-adjust: none;
color: WindowText; /* Should be CanvasText */
```

### Participation

- [ ] I am willing to submit a pull request to implement this rule.

### Additional comments

Additional info: https://blogs.windows.com/msedgedev/2020/09/17/styling-for-windows-high-contrast-with-new-standards-for-forced-colors/

**Comment by nzakas** [comment]
Thanks for the suggestion. There really isn't enough information in the description to understand what you're suggesting or why. Can you add more details?

**Comment by NotWoods** [comment]
Sure thing. `-ms-high-contrast` refers to a family of properties, keywords, and media rules that were used to style in high contrast mode prior to `forced-colors` getting standardized. Due to how many different niches there are, I felt it made more sense to have it as a dedicated rule rather than a blanket no-vendor-prefixes rule (primarily for the color keywords I'll outline below).

# System color keywords
[System colors](https://www.w3.org/TR/css-color-4/#css-system-colors), similar to named colors like `rebeccapurple`, can be used anywhere a color value is supported in CSS.
```css
a {
  color: LinkText;
}
```
Various system colors were used with `-ms-high-contrast` and have been replaced.
Theme color in Windows high contrast settings | Internet Explorer and Microsoft Edge Legacy | New Microsoft Edge and web standards
-- | -- | --
Text | WindowText | CanvasText
Hyperlinks | -ms-hotlight | LinkText
Disabled Text | GrayText
Selected Text | HighlightText (foreground), Highlight (background) | HighlightText (foreground), Highlight (background)
Button Text | ButtonText (foreground), ButtonFace (background) | ButtonText (foreground), ButtonFace (background)
Background | Window | Canvas

The linter rule should ban these values and autofix to the modern equivalent. A possible enhancement is to allow the deprecated values as fallbacks (maybe using a rule option like `allowFallback`).
## Before
```css
a {
  color: -ms-hotlight;
}
```
## After
```css
a {
  color: LinkText;
}

/* or with allowFallback */
a {
  color: -ms-hotlight;
  color: LinkText;
}
```

**Going through the docs to find this table, I stumbled on [additional deprecated CSS system colors](https://drafts.csswg.org/css-color-4/#deprecated-system-colors). So maybe this should be its own dedicated `no-deprecated-system-colors` rule instead.**

# Property
This one is fairly straightforward. The [`-ms-high-contrast-adjust`](https://learn.microsoft.com/en-us/previous-versions/hh771863(v=vs.85)) property was standardized into [`forced-colors-adjust`](https://developer.mozilla.org/en-US/docs/Web/CSS/forced-color-adjust). Both use `auto` and `none` as the possible values.
## Before
```css
img {
  -ms-high-contrast-adjust: none;
}
```
## After
```css
img {
  forced-color-adjust: none;
}

/* or with allowFallback */
img {
  -ms-high-contrast-adjust: none;
  forced-color-adjust: none;
}
```

# Media query
The `-ms-high-contast` query has been split into `forced-colors` and `prefers-color-scheme`.

Internet Explorer and legacy Microsoft Edge | New Microsoft Edge and web standards
-- | --
`@media (-ms-high-contrast: active) {}` | `@media (forced-colors: active) {}`
`@media (-ms-high-contrast: black-on-white) {}` | `@media (forced-colors: active) and (prefers-color-scheme: light) {}`
`@media (-ms-high-contrast: white-on-black) {}` | `@media (forced-colors: active) and (prefers-color-scheme: dark) {}`

There is also a [prefers-contrast](https://www.w3.org/TR/mediaqueries-5/#prefers-contrast) media query but I haven't seen it advocated for much.

**Comment by nzakas** [comment]
Thanks for the additional detail, this is very helpful. I'm curious how long ago `-ms-high-contrast` was superseded? Just weighing the value of having a rule that is only picking up older, outdated things that people may not run into that much.

**Comment by NotWoods** [comment]
For what it's worth, I've had to catch this manually in 5+ PRs over the past year. (But I also work at Microsoft where we're more likely to support high contrast mode.) prefers-forced-colors was added in Chrome 89 three years ago.

**Comment by nzakas** [comment]
> To be more generalized, this proposal could be split into no-deprecated-system-colors and no-vendor-prefixes. Both of these possible rules can have suggestions or autofixes to use the standardized alternatives.

`no-deprecated-system-colors` - I also wonder if this would be too specific. Are there other values besides system colors that have been deprecated? Would it better to have a `no-deprecated-values` rule?

`no-vendor-prefixes` - I'm not quite sure what this might do. The name sounds like it would ban all vendor prefixes, but from the context of this conversation, it seems like maybe your intent was to flag properties where vendor prefixes are no longer needed?


And keep in mind, just like JavaScript rules, you can always create your own custom rules to do whatever you want. My goal here is to ensure this plugin has rules that are as widely-applicable as possible.

**Comment by armano2** [comment]
> `no-vendor-prefixes` - I'm not quite sure what this might do. The name sounds like it would ban all vendor prefixes, but from the context of this conversation, it seems like maybe your intent was to flag properties where vendor prefixes are no longer needed?

this would actually be a great rule, witch would play really nicely with tools like [autoprefixer](https://github.com/postcss/autoprefixer) or [lightningcss](https://lightningcss.dev/transpilation.html#vendor-prefixing) to automatically add vendor specific prefixes to desired targets

**Comment by JoshuaKGoldberg** [comment]
> seems like maybe your intent was to flag properties where vendor prefixes are no longer needed?

I also think this would be a great rule. I think there are generally three classifications of vendor properties:

1. Old ones: no longer needed, should ideally be linted away. Example: [`-ms-high-contrast`](https://blogs.windows.com/msedgedev/2024/04/29/deprecating-ms-high-contrast/).
2. Current ones: in common use for some projects, but won't be forever. Example: [`-webkit-lime-clamp`](https://developer.mozilla.org/en-US/docs/Web/CSS/line-clamp)
3. Unknown/upcoming ones: such as those that are early stage and are not ratified. Example: [`-moz-orient`](https://developer.mozilla.org/en-US/docs/Web/CSS/-moz-orient)

How about `no-deprecated-vendor-prefixes` as a rule name? That would reflect that it's a more specific equivalent to the general [`@typescript-eslint/no-deprecated`](https://typescript-eslint.io/rules/no-deprecated).

A separate rule to ban _non-standard_ prefixes (classification 3) might be a good separate discussion. But I don't think it should be in scope for this rule around _deprecated_ prefixes.
